src/compiler/resolutionCache.ts

```diff
@@ -0,0 +1,316 @@
+            filesWithChangedSetOfUnresolvedImports = undefined;
+            return collected;
+        }
+
+        function createHasInvalidatedResolution(): HasInvalidatedResolution {
+            const collected = filesWithInvalidatedResolutions;
+            filesWithInvalidatedResolutions = undefined;
+            return path => collected && collected.has(path);
+        }
+
+        function resolveModuleName(moduleName: string, containingFile: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost): ResolvedModuleWithFailedLookupLocations {
+            const primaryResult = ts.resolveModuleName(moduleName, containingFile, compilerOptions, host);
+            // return result immediately only if it is .ts, .tsx or .d.ts
+            // otherwise try to load typings from @types
+            return (resolveWithGlobalCache && resolveWithGlobalCache(primaryResult, moduleName, compilerOptions, host)) || primaryResult;
+        }
+
+        function resolveNamesWithLocalCache<T extends NameResolutionWithFailedLookupLocations, R>(
+            names: string[],
+            containingFile: string,
+            cache: Map<Map<T>>,
+            loader: (name: string, containingFile: string, options: CompilerOptions, host: ModuleResolutionHost) => T,
+            getResult: (s: T) => R,
+            getResultFileName: (result: R) => string | undefined,
+            logChanges: boolean): R[] {
+
+            const path = toPath(containingFile);
+            const currentResolutionsInFile = cache.get(path);
+
+            const newResolutions: Map<T> = createMap<T>();
+            const resolvedModules: R[] = [];
+            const compilerOptions = getCompilerOptions();
+
+            for (const name of names) {
+                // check if this is a duplicate entry in the list
+                let resolution = newResolutions.get(name);
+                if (!resolution) {
+                    const existingResolution = currentResolutionsInFile && currentResolutionsInFile.get(name);
+                    if (moduleResolutionIsValid(existingResolution)) {
+                        // ok, it is safe to use existing name resolution results
+                        resolution = existingResolution;
+                    }
+                    else {
+                        resolution = loader(name, containingFile, compilerOptions, host);
+                        updateFailedLookupLocationWatches(containingFile, name, existingResolution && existingResolution.failedLookupLocations, resolution.failedLookupLocations);
+                    }
+                    newResolutions.set(name, resolution);
+                    if (logChanges && filesWithChangedSetOfUnresolvedImports && !resolutionIsEqualTo(existingResolution, resolution)) {
+                        filesWithChangedSetOfUnresolvedImports.push(path);
+                        // reset log changes to avoid recording the same file multiple times
+                        logChanges = false;

```

Comment:
The resolution is per file and we are recording the files so anyways we need to record only once. This is same stuff earlier in lsHost thats moved as a helper